x/marker/handler_test.go

```diff
@@ -72,58 +81,768 @@ func TestInvalidProposal(t *testing.T) {
+			[]string{s.user1},
+			"",
+			nil,
+			0,
+		},
+		{
+			"setup grant access to marker",
+			types.NewMsgAddAccessRequest(hotdogDenom, s.user1Addr, accessMintGrant),
+			[]string{s.user1},
+			"",
+			nil,
+			0,
+		},
+		{
+			"should successfully delete grant access to marker",
+			types.NewDeleteAccessRequest(hotdogDenom, s.user1Addr, s.user1Addr),
+			[]string{s.user1},
+			"",
+			types.NewEventMarkerDeleteAccess(s.user1, hotdogDenom, s.user1),
+			2,
+		},
+	}
+	for _, tc := range cases {
+		s.T().Run(tc.name, func(t *testing.T) {
+			_, err := s.handler(s.ctx, tc.msg)
+			if len(tc.errorMsg) > 0 {
+				assert.EqualError(t, err, tc.errorMsg)
+			} else {
+				assert.NoError(t, err)
+				if tc.expectedEvent != nil {
+					em := s.ctx.EventManager()
+					msg1, _ := sdk.ParseTypedEvent(em.Events().ToABCIEvents()[tc.eventIdx])
+					require.Equal(t, tc.expectedEvent, msg1)
+				}
+			}
+		})
+	}
+}
+
+func (s HandlerTestSuite) TestMsgFinalizeMarkerRequest() {
+
+	hotdogDenom := "hotdog"
+
+	cases := []struct {
+		name          string
+		msg           sdk.Msg
+		signers       []string
+		errorMsg      string
+		expectedEvent *types.EventMarkerFinalize
+		eventIdx      int
+	}{
+		{
+			"setup new marker for test",
+			types.NewMsgAddMarkerRequest(hotdogDenom, sdk.NewInt(100), s.user1Addr, s.user1Addr, types.MarkerType_Coin, true, true),
+			[]string{s.user1},
+			"",
+			nil,
+			0,
+		},
+		{
+			"should successfully finalize marker",
+			types.NewMsgFinalizeRequest(hotdogDenom, s.user1Addr),
+			[]string{s.user1},
+			"",
+			types.NewEventMarkerFinalize(hotdogDenom, s.user1),
+			1,
+		},
+	}
+	for _, tc := range cases {
+		s.T().Run(tc.name, func(t *testing.T) {
+			_, err := s.handler(s.ctx, tc.msg)
+			if len(tc.errorMsg) > 0 {
+				assert.EqualError(t, err, tc.errorMsg)
+			} else {
+				assert.NoError(t, err)
+				if tc.expectedEvent != nil {
+					em := s.ctx.EventManager()
+					msg1, _ := sdk.ParseTypedEvent(em.Events().ToABCIEvents()[tc.eventIdx])
+					require.Equal(t, tc.expectedEvent, msg1)
+				}
+			}
+		})
+	}
+}
+
+func (s HandlerTestSuite) TestMsgActivateMarkerRequest() {
+
+	hotdogDenom := "hotdog"
+
+	cases := []struct {
+		name          string
+		msg           sdk.Msg
+		signers       []string
+		errorMsg      string
+		expectedEvent *types.EventMarkerActivate
+		eventIdx      int
+	}{
+		{
+			"setup new marker for test",
+			types.NewMsgAddMarkerRequest(hotdogDenom, sdk.NewInt(100), s.user1Addr, s.user1Addr, types.MarkerType_Coin, true, true),
+			[]string{s.user1},
+			"",
+			nil,
+			0,
+		},
+		{
+			"setup finalize marker",
+			types.NewMsgFinalizeRequest(hotdogDenom, s.user1Addr),
+			[]string{s.user1},
+			"",
+			nil,
+			0,
+		},
+		{
+			"should successfully activate marker",
+			types.NewMsgActivateRequest(hotdogDenom, s.user1Addr),
+			[]string{s.user1},
+			"",
+			types.NewEventMarkerActivate(hotdogDenom, s.user1),
+			4, //finalize marker will emit a send and message event from bank module
```

Comment:
the event index stuff feels kind of brittle... perhaps a helper method that does a "contains event" check instead... otherwise if the bank module (or another future module involved here) changes events these tests will break...

Reply:
Yes, that is even better.  I will do that.  My original idea was to clear it after each test so it would always be the event at index 0, but that isn't possible...